cli/npm/cache.rs

```diff
@@ -0,0 +1,208 @@
+    Self::new(dir.root.join("npm"))
+  }
+
+  pub fn package_folder(
+    &self,
+    id: &NpmPackageId,
+    registry_url: &Url,
+  ) -> PathBuf {
+    self
+      .package_name_folder(&id.name, registry_url)
+      .join(id.version.to_string())
+  }
+
+  pub fn package_name_folder(&self, name: &str, registry_url: &Url) -> PathBuf {
+    let mut dir = self
+      .root_dir
+      .join(fs_util::root_url_to_safe_local_dirname(registry_url));
+    // ensure backslashes are used on windows
+    for part in name.split('/') {
+      dir = dir.join(part);
+    }
+    dir
+  }
+
+  pub fn resolve_package_id_from_specifier(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<NpmPackageId, AnyError> {
+    match self.maybe_resolve_package_id_from_specifier(specifier) {
+      Some(id) => Ok(id),
+      None => bail!("could not find npm package for '{}'", specifier),
+    }
+  }
+
+  fn maybe_resolve_package_id_from_specifier(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<NpmPackageId> {
+    let relative_url = self.root_dir_url.make_relative(specifier)?;
+    if relative_url.starts_with("../") {
+      return None;
+    }
+
+    // examples:
+    // * chalk/5.0.1/
+    // * @types/chalk/5.0.1/
+    let is_scoped_package = relative_url.starts_with('@');
+    let mut parts = relative_url
+      .split('/')
+      .enumerate()
+      .take(if is_scoped_package { 3 } else { 2 })
+      .map(|(_, part)| part)
+      .collect::<Vec<_>>();
+    let version = parts.pop().unwrap();
+    let name = parts.join("/");
+
+    Some(NpmPackageId {
+      name,
+      version: semver::Version::parse(version).unwrap(),
+    })
+  }
+}
+
+/// Stores a single copy of npm packages in a cache.
+#[derive(Clone, Debug)]
+pub struct NpmCache(ReadonlyNpmCache);
+
+impl NpmCache {
+  pub fn new(root_dir: PathBuf) -> Self {
+    Self(ReadonlyNpmCache::new(root_dir))
+  }
+
+  pub fn from_deno_dir(dir: &DenoDir) -> Self {
+    Self(ReadonlyNpmCache::from_deno_dir(dir))
+  }
+
+  pub fn as_readonly(&self) -> ReadonlyNpmCache {
+    self.0.clone()
+  }
+
+  pub async fn ensure_package(
+    &self,
+    id: &NpmPackageId,
+    dist: &NpmPackageVersionDistInfo,
+    registry_url: &Url,
+  ) -> Result<(), AnyError> {
+    let package_folder = self.0.package_folder(id, registry_url);
+    if package_folder.exists()
+      // if this file exists, then the package didn't successfully extract
+      // the first time, or another process is currently extracting the zip file
+      && !package_folder.join(NPM_PACKAGE_SYNC_LOCK_FILENAME).exists()
+    {
+      return Ok(());
+    }
+
+    log::log!(
+      log::Level::Info,
+      "{} {}",
+      colors::green("Download"),
+      dist.tarball,
+    );
+
+    let response = reqwest::get(&dist.tarball).await?;
+
+    if response.status() == 404 {
+      bail!("Could not find npm package tarball at: {}", dist.tarball);
+    } else if !response.status().is_success() {
+      bail!("Bad response: {:?}", response.status());
```

Comment:
It will show the package name at a higher level for these. There's a `.with_context(|| format!("Failed caching npm package '{}'.", package.id))` on the call to ensure_package.